// streams/async-iterator-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncIteratorStream = void 0;
const iterator_1 = require("./sources/iterator");
const as_async_iterable_1 = require("./util/as-async-iterable");
//
// AsyncIteratorStream Implementation
//
/**
 * An asynchronous stream that produces elements of type `T` on demand.
 *
 * This is an extension of the built-in `AsyncIterable<T>` protocol.
 *
 * The operations defined here in AsyncIteratorStream are a superset of the
 * operations described in the
 * [Async Iterator Helpers](https://github.com/tc39/proposal-async-iterator-helpers)
 * proposal.
 *
 * The behavior of all operations here that correspond to operations in the
 * `Async Iterator Helpers` proposal are defined (as best as possible) to match
 * exactly with the behavior of the corresponding `Async Iterator Helpers`
 * operations. This is intended to allow users of this library to seamlessly
 * switch between this library and the `Async Iterator Helpers` once those are
 * implemented and available.
 *
 * However, this library is NOT a polyfill for `Async Iterator Helpers`. To use
 * this library, an async iterable iterator or generator needs to be explicitly
 * wrapped in an AsyncIteratorStream like in this example:
 *
 * ```
 * import "streams/asyncGenerator";
 *
 * async function* generator() {
 *   yield* [1, 2, 3];
 * }
 *
 * generator()               // returns an AsyncGenerator
 *   .stream()               // convert to AsyncIteratorStream first!
 *   .forEach(console.log);  // use the AsyncIteratorStream APIs
 * ```
 */
class AsyncIteratorStream {
    iterator;
    /**
     * Creates an AsyncIteratorStream from an Iterator or Iterable.
     *
     * ```
     * AsyncIteratorStream.from([1, 2, 3]).map(x => x * 2).forEach(console.log);
     * ```
     */
    static from(it) {
        if (typeof it[Symbol.asyncIterator] === "function") {
            return new AsyncIteratorStream(it[Symbol.asyncIterator]());
        }
        if (typeof it[Symbol.iterator] === "function") {
            return new AsyncIteratorStream((0, iterator_1.toAsync)(it[Symbol.iterator]()));
        }
        return new AsyncIteratorStream(it);
    }
    // The AsyncIterator protocol
    next;
    return;
    throw;
    constructor(iterator) {
        this.iterator = iterator;
        this.next = () => this.iterator.next();
        if (this.iterator.return !== undefined) {
            this.return = () => this.iterator.return();
        }
        if (this.iterator.throw !== undefined) {
            this.throw = (e) => this.iterator.throw(e);
        }
    }
    stream() {
        return this;
    }
    [Symbol.asyncIterator]() {
        return this;
    }
    async [Symbol.asyncDispose]() {
        await this.return?.();
    }
    // #region Intermediate operations
    /**
     * Returns a new stream that skips elements of this stream not matched by the
     * `predicate`.
     *
     * See also [IteratorHelpers#filter](https://github.com/tc39/proposal-iterator-helpers#filterfiltererfn).
     *
     * @param predicate a function that decides whether to include each element
     * in the new stream (true) or to exclude the element (false)
     */
    filter(predicate) {
        async function* filterOperator(it) {
            for await (const v of it) {
                if (predicate(v)) {
                    yield v;
                }
            }
        }
        return new AsyncIteratorStream(filterOperator(this));
    }
    /**
     * Returns a new stream that transforms each element of this stream
     * using the provided function.
     *
     * See also [IteratorHelpers#map](https://github.com/tc39/proposal-iterator-helpers#mapmapperfn).
     *
     * @param transform a function to apply to each element of this stream
     */
    map(transform) {
        // Do not use a generator here, as it will await the result of the
        // transform before yielding each mapped value.
        // The Iterator Helpers proposal requires that we yield immediately
        // without awaiting the transform result to resolve.
        function mapOperator(itrbl) {
            const it = itrbl[Symbol.asyncIterator]();
            function mapResult(sourceResult) {
                return new Promise((resolve, reject) => {
                    Promise.resolve(sourceResult).then((result) => {
                        if (result.done) {
                            resolve({ done: true, value: undefined });
                        }
                        else {
                            // transform could return a Promise...
                            const transformed = transform(result.value);
                            Promise.resolve(transformed).then((value) => resolve({ value, done: false }), reject);
                        }
                    }, reject);
                });
            }
            const nextMapped = () => {
                const sourceResult = it.next();
                return mapResult(sourceResult);
            };
            const mappedIterator = { next: nextMapped };
            if (it.return) {
                mappedIterator.return = (value) => {
                    const sourceResult = it.return(value);
                    return mapResult(sourceResult);
                };
            }
            if (it.throw) {
                mappedIterator.throw = (e) => {
                    const sourceResult = it.throw(e);
                    return mapResult(sourceResult);
                };
            }
            return mappedIterator;
        }
        return new AsyncIteratorStream(mapOperator(this));
    }
    /**
     * Like `map` but the result from each call to `transform` is awaited
     * before producing the next transformed element.
     *
     * @param transform an async function to apply to each element of this stream
     */
    mapAwait(transform) {
        async function* mapAwaitOperator(it) {
            for await (const v of it) {
                yield transform(v);
            }
        }
        return new AsyncIteratorStream(mapAwaitOperator(this));
    }
    /**
     *
     * See also [IteratorHelpers#flatMap](https://github.com/tc39/proposal-iterator-helpers#flatmapmapperfn).
     *
     * @param transform
     */
    flatMap(transform) {
        async function* flatMapOperator(it) {
            for await (const nested of it) {
                yield* (0, as_async_iterable_1.asAsyncIterable)(transform(nested));
            }
        }
        return new AsyncIteratorStream(flatMapOperator(this));
    }
    batch(batchSize) {
        if (batchSize < 1) {
            throw new Error("batchSize should be positive");
        }
        async function* batchOperator(it) {
            let acc = [];
            for await (const v of it) {
                acc.push(v);
                if (acc.length === batchSize) {
                    yield acc;
                    acc = [];
                }
            }
            if (acc.length > 0) {
                yield acc;
            }
        }
        return new AsyncIteratorStream(batchOperator(this));
    }
    /**
     * Returns a new stream that produces up to the first `limit` number of
     * elements of this stream.
     *
     * See also [IteratorHelpers#take](https://github.com/tc39/proposal-iterator-helpers#takelimit).
     *
     * @param limit the maximum number of items to produce
     */
    take(limit) {
        async function* takeOperator(it) {
            let count = 0;
            if (count >= limit) {
                return;
            }
            for await (const v of it) {
                yield v;
                count += 1;
                if (count >= limit) {
                    return;
                }
            }
        }
        return new AsyncIteratorStream(takeOperator(this));
    }
    /**
     *
     * See also [IteratorHelpers#drop](https://github.com/tc39/proposal-iterator-helpers#droplimit).
     *
     * @param n
     */
    drop(n) {
        async function* dropOperator(it) {
            let count = 0;
            for await (const v of it) {
                if (count >= n) {
                    yield v;
                }
                count += 1;
            }
        }
        return new AsyncIteratorStream(dropOperator(this));
    }
    dropWhile(predicate) {
        async function* dropWhileOperator(it) {
            let dropping = true;
            for await (const v of it) {
                dropping = dropping && predicate(v);
                if (!dropping) {
                    yield v;
                }
            }
        }
        return new AsyncIteratorStream(dropWhileOperator(this));
    }
    takeWhile(predicate) {
        async function* takeWhileOperator(it) {
            for await (const v of it) {
                if (!predicate(v)) {
                    return;
                }
                yield v;
            }
        }
        return new AsyncIteratorStream(takeWhileOperator(this));
    }
    peek(observer) {
        async function* peekOperator(it) {
            for await (const v of it) {
                observer(v);
                yield v;
            }
        }
        return new AsyncIteratorStream(peekOperator(this));
    }
    //
    // Terminal operations
    //
    /**
     *
     * See also [IteratorHelpers#forEach](https://github.com/tc39/proposal-iterator-helpers#foreachfn).
     *
     * @param block
     */
    async forEach(block) {
        for await (const v of this) {
            await block(v);
        }
    }
    async collect(container, accumulator, finisher) {
        for await (const v of this) {
            accumulator(container, v);
        }
        return finisher ? finisher(container) : container;
    }
    /**
     *
     * See also [IteratorHelpers#every](https://github.com/tc39/proposal-iterator-helpers#everyfn).
     *
     * @param predicate
     */
    async every(predicate) {
        for await (const v of this) {
            if (!(await predicate(v))) {
                return false;
            }
        }
        return true;
    }
    async some(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return true;
            }
        }
        return false;
    }
    async none(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return false;
            }
        }
        return true;
    }
    async count() {
        let count = 0;
        for await (const _ of this) {
            count += 1;
        }
        return count;
    }
    /**
     * Returns the first element that matches the predicate.
     *
     * This is the same as the {@link first()} method except that the predicate is
     * required and a `TypeError` will be thrown if a predicate is not supplied.
     *
     * See also [IteratorHelpers#find](https://github.com/tc39/proposal-iterator-helpers#findfn).
     *
     * @param predicate
     */
    async find(predicate) {
        for await (const v of this) {
            if (await predicate(v)) {
                return v;
            }
        }
        return undefined;
    }
    async first(predicate = (_) => true) {
        for await (const v of this) {
            if (await predicate(v)) {
                return v;
            }
        }
        return undefined;
    }
    async last(predicate = (_) => true) {
        let result;
        for await (const v of this) {
            if (await predicate(v)) {
                result = v;
            }
        }
        return result;
    }
    async max(comparator) {
        let result;
        let firstItem = true;
        for await (const v of this) {
            if (firstItem) {
                result = v;
                firstItem = false;
            }
            else {
                result = comparator(result, v) > 0 ? result : v;
            }
        }
        return result;
    }
    async min(comparator) {
        let result;
        let firstItem = true;
        for await (const v of this) {
            if (firstItem) {
                result = v;
                firstItem = false;
            }
            else {
                result = comparator(result, v) < 0 ? result : v;
            }
        }
        return result;
    }
    /**
     *
     * See also [IteratorHelpers#reduce](https://github.com/tc39/proposal-iterator-helpers#reducereducer--initialvalue-).
     *
     * @param reducer
     * @param initial
     */
    async reduce(reducer, initial) {
        const hasInitial = arguments.length >= 2;
        let firstItem = !hasInitial;
        let result = initial;
        for await (const v of this) {
            if (firstItem) {
                result = v; // R is assumed to be T when there is no initial value
                firstItem = false;
            }
            else {
                result = reducer(result, v);
            }
        }
        if (firstItem) {
            throw new TypeError("reduce without initial value but stream is empty");
        }
        return result;
    }
    async fold(reducer, initial) {
        const hasInitial = arguments.length >= 2;
        let firstItem = !hasInitial;
        let result = initial;
        for await (const v of this) {
            if (firstItem) {
                result = v; // R is assumed to be T when there is no initial value
                firstItem = false;
            }
            else {
                result = reducer(result, v);
            }
        }
        return result;
    }
    /**
     * See also [IteratorHelpers#toArray](https://github.com/tc39/proposal-iterator-helpers#toarray).
     */
    async toArray() {
        const result = [];
        for await (const v of this) {
            result.push(v);
        }
        return result;
    }
}
exports.AsyncIteratorStream = AsyncIteratorStream;
