// streams/sources/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromEvent = fromEvent;
exports.fromEventPattern = fromEventPattern;
const async_iterator_pair_1 = require("../util/async-iterator-pair");
function fromEvent(target, eventName, options) {
    if (typeof target.on === "function") {
        return fromEventPattern((handler) => target.on(eventName, handler), (handler) => target.off?.(eventName, handler));
    }
    else if (typeof target.addListener === "function") {
        return fromEventPattern((handler) => target.addListener(eventName, handler), (handler) => target.removeListener?.(eventName, handler));
    }
    else if (typeof target.addEventListener === "function") {
        return fromEventPattern((handler) => target.addEventListener(eventName, handler, options), (handler) => target.removeEventListener?.(eventName, handler, options));
    }
    throw new TypeError("unsupported event target");
}
function fromEventPattern(addHandler, removeHandler) {
    const [eventConsumer, eventProducer] = (0, async_iterator_pair_1.makeAsyncGeneratorPair)({
        onReturn: stop,
    });
    const eventHandler = (value) => eventProducer.next(value);
    function start() {
        addHandler(eventHandler);
    }
    function stop() {
        removeHandler?.(eventHandler);
    }
    async function* eventGenerator() {
        start();
        return yield* eventConsumer;
    }
    return eventGenerator();
}
